torchx/components/dist.py

```diff
@@ -80,6 +86,293 @@
+        specified, then it is honored even if it does not match the number of GPUs on the host.
+        However, a warning message is displayed reminding the user that there is a mismatch between
+        the GPU count on the host and the configured nproc_per_node.
+
+        Example (GPU):
+
+            .. doctest::
+
+                >>> from torchx.components.dist import ArgJ
+                >>> str(ArgJ.parse_from(h="aws_p4d.24xlarge", j="2"))
+                '2x8'
+
+                >>> str(ArgJ.parse_from(h="aws_p4d.24xlarge", j="2x4"))
+                '2x4'
+
+
+        For hosts with no GPU devices, one MUST specify nproc_per_node. Otherwise this function will
+        raise an Error.
+
+        Example (CPU or Trainium):
+
+            .. doctest::
+
+                >>> str(ArgJ.parse_from(h="aws_trn1.32xl", j="2"))
+                Traceback (most recent call last):
+                    ...
+                ValueError: nproc_per_node cannot be inferred from GPU count. `trn1.32xl` is not a GPU instance. ...
+
+                >>> str(ArgJ.parse_from(h="aws_trn1.32xl", j="2x16"))
+                '2x16'
+
+        """
+        nums = j.split("x")
+        num_gpus = specs.named_resources[h].gpu
+        if len(nums) == 1:  # -j 1
+            nnodes = int(nums[0])
+            # infer nproc_per_node from # of gpus in host
+
+            if num_gpus > 0:
+                nproc_per_node = num_gpus
+            else:
+                raise ValueError(
+                    f"nproc_per_node cannot be inferred from GPU count."
+                    f" `{h}` is not a GPU instance."
+                    f" You must specify `-j $NNODESx$NPROCS_PER_NODE` (e.g. `-j {nnodes}x8`)"
+                )
+
+        elif len(nums) == 2:  # -j 1x2
+            nnodes = int(nums[0])
+            nproc_per_node = int(nums[1])
+
+            if nproc_per_node != num_gpus:
+                warnings.warn(
+                    f"In `-j {j}` you specified nproc_per_node={nproc_per_node}"
+                    f" which does not equal the number of GPUs on a {h}: {num_gpus}."
+                    f" This may lead to under-utilization or an error. "
+                    f" If this was intentional, ignore this warning."
+                    f" Otherwise set `-j {nnodes}` to auto-set nproc_per_node"
+                    f" to the number of GPUs on the host."
+                )
+        else:
+            raise ValueError(
+                f"Invalid format for `-j $NNODESx$NPROCS_PER_NODE` (e.g. `-j 1x8`). Given: {j}"
+            )
+
+        return ArgJ(nnodes=nnodes, nproc_per_node=nproc_per_node)
+
+
+def default_experiment_name() -> str:
+    return f"default-experiment/{getuser()}/{socket.getfqdn()}"
+
+
+def spmd(
+    *args: str,
+    script: Optional[str] = None,
+    m: Optional[str] = None,
+    image: str = torchx.IMAGE,
+    name: str = "/",
+    h: str = "aws_p3.2xlarge",
```

Comment:
this shouldn't default to aws

Reply:
created generic named resources and defaulted this to `gpu.small`